lib/tests/streamlit/runtime/caching/singleton_test.py

```diff
@@ -70,6 +70,63 @@ def f():
         self.assertEqual(r2, [1, 1])
 
 
+class SingletonValidateTest(unittest.TestCase):
+    def setUp(self) -> None:
+        # Caching functions rely on an active script run ctx
+        add_script_run_ctx(threading.current_thread(), create_mock_script_run_ctx())
+
+    def tearDown(self):
+        st.experimental_singleton.clear()
+        # Some of these tests reach directly into _cache_info and twiddle it.
+        # Reset default values on teardown.
+        singleton_decorator.SINGLETON_CALL_STACK._cached_func_stack = []
+        singleton_decorator.SINGLETON_CALL_STACK._suppress_st_function_warning = 0
+
+    def test_validate_success(self):
+        """If we have a validate function and it returns True, we don't recompute our cached value."""
+        validate = Mock(return_value=True)
+
+        call_count: List[int] = [0]
+
+        @st.experimental_singleton(validate=validate)
+        def f() -> int:
+            call_count[0] += 1
+            return call_count[0]
+
+        # First call: call_count == 1; validate not called (because we computed a new value)
+        self.assertEqual(1, f())
+        validate.assert_not_called()
+
+        # Subsequent calls: call_count == 1; validate called each time
+        for _ in range(3):
+            self.assertEqual(1, f())
+            validate.assert_called_once_with(1)
+            validate.reset_mock()
+
+    def test_validate_fail(self):
+        """If we have a validate function and it returns False, we recompute our cached value."""
+        validate = Mock(return_value=False)
+
+        call_count: List[int] = [0]
+
+        @st.experimental_singleton(validate=validate)
+        def f() -> int:
+            call_count[0] += 1
+            return call_count[0]
+
+        # First call: call_count == 1; validate not called (because we computed a new value)
+        expected_call_count = 1
+        self.assertEqual(expected_call_count, f())
+        validate.assert_not_called()
+
+        # Subsequent calls: call_count increases; validate called with previous value
+        for _ in range(3):
+            expected_call_count += 1
+            self.assertEqual(expected_call_count, f())
+            validate.assert_called_once_with(expected_call_count - 1)
+            validate.reset_mock()
+
+
```

Comment:
@tconkling could we add a test that if the `validate` function always return `False` (something like `lambda x: return False`, an infinite loop of calling validate doesn't happen, and just a new computed value returned

Reply:
@kajarenc That should be what's happening in this test already, I think? (Our `validate` function is a Mock that always returns False, and it's being called multiple times in a loop.) Let me know if I'm misunderstanding, though!

Reply:
Ou, yes, it is :)
I initially did not understand that this test checks this too!

Reply:
(Maybe this is confusion over the `reset_mock` call? That resets the call count on the `Mock` object, but it doesn't reset the `return_value=False`, declared at the top of the test)

Reply:
Ah great!